### ISSUE TITLE
Generic classes with nested generic classes can't have different TypeVars

### ISSUE THREAD

**Issue by euresti** [python/mypy#5551]
Looking at the following example
```
from typing import TypeVar, Generic

class C(Generic[T]):
    def __init__(self, t: T) -> None:
        self.t = t

    class F(Generic[K]):
        def __init__(self, k: K) -> None:
            self.k = k

        def foo(self) -> K:
            return self.k

C.F(17).foo()
C("").F(17).foo()  # error: Argument 1 to "F" has incompatible type "int"; expected "str"

reveal_type(C.F)  # 'def [K] (k: K`1) -> foo.C.F[K`1]'  All Good
reveal_type(C("").F)  # 'def [K] (k: builtins.str*) -> foo.C.F[builtins.str*]'   What?
```
Seems like specifying the `T` seems to crush `K` somehow.

**Comment by euresti** [comment]
I don't have any code that does this.  I was just testing out a fix for #5542 and came across this issue.

**Comment by ilevkivskyi** [comment]
Yes, there are still some weird clashes that can happen. I would say we need to "fix" https://github.com/python/mypy/issues/4814 soon.

**Comment by AlexWaygood** [comment]
@JelleZijlstra, this appears to be another one that was fixed by #12590. Reckon we should add another test case?

**Comment by JelleZijlstra** [comment]
Yes, this seems good to test.
